internal/driver/readhandler.go

```diff
@@ -103,7 +105,26 @@ func makeReadRequest(deviceClient *opcua.Client, req sdkModel.CommandRequest) (*
 
 	// make new result
 	reading := resp.Results[0].Value.Value()
-	return newResult(req, reading)
+	result, err := newResult(req, reading)
+
+	// Try to get the sourcetime timestamp from the result since it is the closest timestamp to the event source. If not applicable, use servertime or the current time.
+	if err == nil {
+		var sourceTime = resp.Results[0].SourceTimestamp.UnixNano() / int64(time.Millisecond)
```

Comment:
I think it's 'cleaner' to check the time at first for being zero - before dividing it or using it in expressions. And [there is a special method of `time.Time` `IsZero`](https://pkg.go.dev/time#pkg-functions) which is intended for this case.

Possibly by inverting the test to keep test and application of one source of time closer together in resulting source code one could write:

```suggestion
        var tm time.Time
	if !resp.Results[0].SourceTimestamp.IsZero() {
		tm = resp.Results[0].SourceTimestamp
	} else if !resp.Results[0].ServerTimestamp.IsZero() {
		tm = resp.Results[0].ServerTimestamp
	} else {
		tm = time.Now()
	}
	result.Origin = tm.UnixNano() / int64(time.Millisecond)
```
The conditional extraction of the variable `tm` from source time, server time or local time should best be extracted into a separate function. The variable `tm` can be omittet this way by `result.Origin = selectOriginTime(resp.Results[0]`.

Naming and structuring above are of course just suggestions.

Reply:
Just to add in something here, if we could avoid if else, that would be my way to go. Having more conditions and paths only reduces the performance.

Reply:
Do you have an example to proceed here without any if else?

Reply:
Done, really not sure how to  remove the conditions here.